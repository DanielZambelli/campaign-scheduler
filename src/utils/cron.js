const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
