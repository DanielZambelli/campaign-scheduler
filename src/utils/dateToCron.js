const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
