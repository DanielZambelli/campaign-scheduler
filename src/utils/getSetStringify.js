const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
