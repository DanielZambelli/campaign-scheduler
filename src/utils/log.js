const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
