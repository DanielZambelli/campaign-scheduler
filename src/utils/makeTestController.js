const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
