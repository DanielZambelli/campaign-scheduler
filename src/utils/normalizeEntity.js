const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
