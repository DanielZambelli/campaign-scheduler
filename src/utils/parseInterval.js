const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
