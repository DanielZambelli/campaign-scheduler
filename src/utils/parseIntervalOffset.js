const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
