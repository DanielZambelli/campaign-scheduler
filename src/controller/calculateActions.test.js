require('../utils/test')
const {seed} = require('../utils/test/seeds1')
const TEST = 'calculate_actions'

describe(TEST, () => {

  beforeAll(init(TEST, seed))

  afterAll(destroy)

  it('validates 1', async () => {
    const res = await ctl.calculateActions().catch(e => e.message)
    expect(res).toEqual('requires campaignId')
  })

  it('validates 2', async () => {
    const res = await ctl.calculateActions(null).catch(e => e.message)
    expect(res).toEqual('requires campaignId')
  })

  it('validates 3', async () => {
    const res = await ctl.calculateActions(0).catch(e => e.message)
    expect(res).toEqual('requires campaignId')
  })

  it('validates 4', async () => {
    const res = await ctl.calculateActions(undefined).catch(e => e.message)
    expect(res).toEqual('requires campaignId')
  })

  it('validates 5', async () => {
    const res = await ctl.calculateActions('NON_EXISTING_CAMPAIGN').catch(e => e.message)
    expect(res).toEqual('campaign not found')
  })

  it('validates 6', async () => {
    const res = await ctl.calculateActions('campaign1').then(normalize)
    expect(res.result.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('validates 7', async () => {
    const res = await ctl.calculateActions('campaign2').then(normalize)
    expect(res.result.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('validates 8', async () => {
    const res = await ctl.calculateActions('campaign3').then(normalize)
    expect(res.result.length).toEqual(2)
    expect(res).toMatchSnapshot()
  })

  it('validates 9', async () => {
    let res = await ctl.calculateActions('campaign3','contact#1').then(normalize)
    res = res.result.filter(o => o.subject)
    expect(res.length).toEqual(2)
    expect(res).toMatchSnapshot()
  })

  it('validates 10', async () => {
    await ctl.define({ id: 'campaign4', actions: [] })
    const res = await ctl.calculateActions('campaign4').then(normalize)
    expect(res.result.length).toEqual(0)
    expect(res).toMatchSnapshot()
  })

})
