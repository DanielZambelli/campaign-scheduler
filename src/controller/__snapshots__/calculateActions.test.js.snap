// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate_actions validates 6 1`] = `
{
  "hash": "da81a8c8dcb8b30c36ae4dbcdf6ee48eab766e76",
  "result": [
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign1",
      "expectedAt": "in an hour",
    },
  ],
}
`;

exports[`calculate_actions validates 7 1`] = `
{
  "hash": "585c647699ac9c9425d3d7fef95b73f5ac7a9e13",
  "result": [
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign2",
      "expectedAt": "in an hour",
    },
  ],
}
`;

exports[`calculate_actions validates 8 1`] = `
{
  "hash": "c60aaecd927a4fccaf38362dda06e32a97d9fb3e",
  "result": [
    {
      "actionId": "action1",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign3",
      "expectedAt": "in a day",
    },
    {
      "actionId": "action2",
      "callbackId": "sendEmail",
      "callbackOpts": {
        "view": "emailTemplate1",
      },
      "campaignId": "campaign3",
      "expectedAt": "in 2 days",
    },
  ],
}
`;

exports[`calculate_actions validates 9 1`] = `
[
  {
    "actionId": "action1",
    "callbackId": "sendEmail",
    "callbackOpts": {
      "view": "emailTemplate1",
    },
    "campaignId": "campaign3",
    "expectedAt": "in a day",
    "subject": "contact#1",
  },
  {
    "actionId": "action2",
    "callbackId": "sendEmail",
    "callbackOpts": {
      "view": "emailTemplate1",
    },
    "campaignId": "campaign3",
    "expectedAt": "in 2 days",
    "subject": "contact#1",
  },
]
`;

exports[`calculate_actions validates 10 1`] = `
{
  "hash": "989db2448f309bfdd99b513f37c84b8f5794d2b5",
  "result": [],
}
`;
